Split classes to handle different kind of packages

ycointe opened the pull request:
**Component**: Buildchain, package

<!-- E.g. 'salt', 'containers', 'kubernetes', 'build', 'tests'... -->

**Context**: Metalk8s on Ubuntu

**Summary**: We need to to split some classes in the buildchain to integrate Debian and build its packages.

**Acceptance criteria**: 
- classes separated
- Build still available 
- Centos cluster deployment available


---

<!-- Declare one or more issues to close once this PR gets merged -->

Closes: #1472 

<!-- If you want to refer to an issue while not closing it, use:

See: #ISSUE_NUMBER

-->

ycointe committed:
build/package: split the Package class

The current Package class is actually only about RPM packages. Since we
are gonna add the support for DEB packages, it's now time to split this
class into two:
- an abstract base class Package that provides the common features of a
  package
- a concrete class RPMPackage, that inherits from Package, that provides
  a way to build RPM packages (and thus contains RPM-specific code).

Coming in a future commit: a class DEBPackage to build DEB packages.

Refs: #1472

bert-e commented:
# Hello ycointe,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/metalk8s/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Branches have diverged



This pull request's source branch `improvement/1472-calico-cni-package-creation` has diverged from
`development/2.4` by more than 50 commits.

To avoid any integration risks, please re-synchronize them using one of the
following solutions:

* Merge `origin/development/2.4` into `improvement/1472-calico-cni-package-creation`
* Rebase `improvement/1472-calico-cni-package-creation` onto `origin/development/2.4`

Note: If you choose to rebase, you may have to ask me to rebuild
integration branches using the `reset` command.

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @teddyandrieux

* @monpote

* @jbertran

* @sayf-eddine-scality

* @nicolast

* @cmonfort

* @slaperche-scality

* @thomasdanan

* @zempashi

* @lucieleonard

* @gdemonet

* @carlito-scality

* @nootal

* @alexandre-allard-scality

ycointe committed:
build/repo: split the Repository class

As for packages, the Repository class is about CentOS repository (a
repository of RPM packages). As we are going to support Ubuntu
repository (a repository of DEB packages) we have to split this class
into two:
- a Repository abstract class providing package-independent features
- an RPMRepository providing a way to create RPM-based repository.

Coming soon: a DEBRepository class to create packages repository for
Ubuntu.

Refs: #1472

ycointe committed:
build/docker: remove RPM-specific code from DockerRun

Since we were only building RPM packages, we've made the questionable
decision to move some repeated RPM-specific code directly into the
DockerRun class.
That will no longer works as we will now reuse this class to run a
Debian builder container, so let's move this code outside.

Refs: #1472

ycointe committed:
package/deb: fix sos-report sources path

The source files for sos-report lives in a common directory (because
they are shared with the RPM packages), not in the DEB-specific tree.

Refs: #1472

ycointe committed:
build/package: add an entrypoint for the Debian builder

This commit add a Shell script that will be used by the Debian builder
container to build DEB packages.
It supports:
- the creation of a DEB package from an RPM one, using alien
- the creation of a DEB package from the sources, using debuild

Note that we specify the version in the changelog instead of inside the
control file because counterintuitively `dpkg-buildpackage` infers the
package version number from this file, and not from a "Version" field in
control.

Refs: #1472

bert-e commented:
# Branches have diverged



This pull request's source branch `improvement/1472-calico-cni-package-creation` has diverged from
`development/2.4` by more than 50 commits.

To avoid any integration risks, please re-synchronize them using one of the
following solutions:

* Merge `origin/development/2.4` into `improvement/1472-calico-cni-package-creation`
* Rebase `improvement/1472-calico-cni-package-creation` onto `origin/development/2.4`

Note: If you choose to rebase, you may have to ask me to rebuild
integration branches using the `reset` command.

alexandre-allard commented on packages/debian/entrypoint.sh:
```diff
@@ -0,0 +1,31 @@
+
+rpm2deb() {
+	set -x
+	chown build:build /home/build
```
Why chowning the home directory, the ownership should already be right, no ?

ycointe replied:
Indeed, It was more a kind of "security" to be sur. I delete it.

alexandre-allard commented on packages/debian/entrypoint.sh:
```diff
@@ -0,0 +1,31 @@
+	#need to be done with user "build"
+	set -x
+	chown build:build /home/build
+        pushd /debbuild/PKG 1>/dev/null
```
Why moving in this directory since the line just after you use a full path and right after you move again ?
For me, this line is useless, must be a leftover of a copy/paste from your function above.

alexandre-allard commented on packages/debian/entrypoint.sh:
```diff
@@ -0,0 +1,31 @@
+	set -x
+	chown build:build /home/build
+	pushd /debbuild/PKG 1>/dev/null
+	for RPM in /rpmbuild/SRC/calico*.rpm; do
```
Should be based on a list of packages to alienize, not hardcoded.

alexandre-allard commented on packages/debian/entrypoint.sh:
```diff
@@ -0,0 +1,31 @@
+		pushd /debbuild/PKG/metalk8s-sosreport/debian
+		debuild -b -uc -us
+	done
+	cp -p /debbuild/SRC/* /debbuild/PKG
```
Not really an error, but is there any good reason for using `-p` ? (just for my personal understanding)

ycointe replied:
It was to "preserve" rights on its but no, it was not very needed because of the `chown` applied after.

alexandre-allard commented on packages/debian/entrypoint.sh:
```diff
@@ -0,0 +1,31 @@
+	set -x
+	chown build:build /home/build
+        pushd /debbuild/PKG 1>/dev/null
+	for DEBSRC in /debbuild/PKG; do
```
Will not work, you need to use a wildcard:
```
for debsrc in /debbuild/PKG/*; do
```
And conventionally, you use lowercased variable when the purpose of the variable is local.
Uppercase is reserved for global variables.

alexandre-allard commented on packages/debian/entrypoint.sh:
```diff
@@ -0,0 +1,31 @@
+	chown build:build /home/build
+        pushd /debbuild/PKG 1>/dev/null
+	for DEBSRC in /debbuild/PKG; do
+		pushd /debbuild/PKG/metalk8s-sosreport/debian
```
Hardcoded package name, you're not using the variable set before `$DEBSRC`

alexandre-allard commented on packages/debian/entrypoint.sh:
```diff
@@ -0,0 +1,31 @@
+builddeb() {
+	#need to be done with user "build"
+	set -x
+	chown build:build /home/build
```
Same as above

alexandre-allard commented on packages/debian/entrypoint.sh:
```diff
@@ -0,0 +1,31 @@
+        pushd /debbuild/PKG 1>/dev/null
+	for DEBSRC in /debbuild/PKG; do
+		pushd /debbuild/PKG/metalk8s-sosreport/debian
+		debuild -b -uc -us
```
Need to be run as build user:
```
su build -c "debuild -b -uc -us"
```

ycointe replied:
Not necessary, i prefer to use a `chown` after which will apply another owner that `root`.

alexandre-allard commented on packages/debian/entrypoint.sh:
```diff
@@ -0,0 +1,31 @@
+	for RPM in /rpmbuild/SRC/calico*.rpm; do
+		alien --bump=0 "$RPM"
+	done
+	cp -p /rpmbuild/SRC/*.deb /debbuild/PKG
```
Useless as you already moved to `/debbuild/PKG` before your loop and the result files (*.deb) from alien are generated in the current directory.

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @teddyandrieux

* @monpote

* @jbertran

* @sayf-eddine-scality

* @nicolast

* @cmonfort

* @slaperche-scality

* @thomasdanan

* @zempashi

* @lucieleonard

* @gdemonet

* @carlito-scality

* @nootal

* @alexandre-allard-scality

ycointe committed:
build/repository: make `get_rpm_path` public

We will need to provide the RPM path to the DEBPackage class, so let's
make this helper public

Refs: #1472

ycointe committed:
build/package: add DEBPackage class

Add a new class DEBPackage which permits to build deb packages either
from sources or from RPM package.

To create a new DEB package, you need to specify its source files. This
can be either:
- the path to a RPM file: in such case we will use `alien` to build the
  package
- the path to a directory containing the source files to packages: in
  such case we will use `debuild`

Refs: #1472

ycointe committed:
build/packaging: add task to build the DEB packages

In order to build the required DEB packages, we create a new task
generator, `_build_deb_packages`, that will create a sub-task for each
DEB package to build.

This task relies a a list of pre-defined packages, DEB_TO_BUILD.

Given that some DEB packages depends on the RPM one, we extract into
dedicated constants what is shared between DEB and RPM build tasks:
namely the calico package and the scality repository.

Refs: #1472

slaperche-scality committed:
package/debian: fix control for sos-report

We replace `extra` by `optional` (because `extra` is deprecated)[1].

We also fix the `Standards-Version` value: it's not the version of the
package but the version of the standards (the policy manual and
associated texts) with which the package complies[2].

As for the `Section` value, we use the same as one used by the
`sosreport` package: after all, it's an administration tool.

[1]: https://www.debian.org/doc/debian-policy/ch-archive.html#s-priorities
[2]: https://www.debian.org/doc/debian-policy/ch-controlfields.html#s-f-standards-version

Refs: #1472
Signed-off-by: Sylvain Laperche <sylvain.laperche@scality.com>

ycointe committed:
build/packaging: Remove auto generated file

During the Debian package generation tests, a file named `file` has
been created. This file should not appear in the list of the source
files and will be not because the volume which will be mounted in the
container to build the Debian package is on `read-only`, so this file
can be remove.

Refs: #1472

jbertran commented on buildchain/buildchain/targets/package.py:
```diff
@@ -319,3 +320,114 @@ def _file_from_url(url: str) -> str:
+         )
+        self._sources = sources
+
+    sources = property(operator.attrgetter('_sources'))
```
I think that with a little work the `sources` management can me mutualized into the `Package` class. This can be an improvement ticket I guess.

E: ~~most likely we want to tend towards the current RPM behaviour, where it would be easy to specify 2 different source paths whereas this would be a problem in the current DEB behaviour.~~ after discussion with @slaperche-scality , just create a debt ticket for reworking/rewording the `sources` thing - it turns out I didn't understand it fully but we still have an issue with it.

ycointe replied:
I create a ticket, you can follow the issue REF #1623

jbertran commented on buildchain/buildchain/targets/package.py:
```diff
@@ -319,3 +320,114 @@ def _file_from_url(url: str) -> str:
+            tasks.append(self.build_deb_package())
+        return tasks
+
+    def build_deb_package(self) -> types.TaskDict:
```
Can be renamed `build_package` like the `RPMPackage` equivalent.

jbertran reviewed (commented):
Dayum, dayum, dayum.

LGTM. Great work!

alexandre-allard commented on packages/debian/entrypoint.sh:
```diff
@@ -0,0 +1,52 @@
+    pushd /debbuild/build-pkg 1>/dev/null
+    debuild -b -uc -us
+
+    cp -r /debbuild/*.deb /debbuild/result
```
Not sure you need `-r`, except if you're trying to copy directories named `*.deb`. :)

ycointe replied:
nop, indeed! :) 
Thanks

alexandre-allard commented on packages/debian/entrypoint.sh:
```diff
@@ -0,0 +1,52 @@
+
+    alien --bump=0 /rpmbuild/source.rpm
+
+    cp /*.deb /debbuild/result
```
And what about:
```
pushd /debbuild/result 1> /dev/null
alien --bump=0 /rpmbuild/source.rpm
chown "$TARGET_UID:$TARGET_GID" *
```

alexandre-allard commented on packages/debian/entrypoint.sh:
```diff
@@ -0,0 +1,52 @@
+}
+
+
+case ${1:-''} in
```
Details: Useless quotes, but it does not hurt, I think. :)

ycointe replied:
Change for Debian and RedHat entrypoint :)

ycointe committed:
build/package: clean Debian and RedHat entrypoint

After we add the Debian entrypoint we want to clean
some typo that can be improve for a better readability.

Refs: #1472

ycointe commented:
@bert-e approve

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @teddyandrieux

* @monpote

* @jbertran

* @sayf-eddine-scality

* @nicolast

* @cmonfort

* @slaperche-scality

* @thomasdanan

* @zempashi

* @lucieleonard

* @gdemonet

* @carlito-scality

* @nootal

* @alexandre-allard-scality










*The following options are set:* **approve**

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/2.4`



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/2.4`



The following branches have **NOT** changed:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`



Please check the status of the associated issue None.

Goodbye ycointe.
